docs/dev/design/new-notifications-system.rst

```diff
@@ -0,0 +1,459 @@
+We had pretty similar ideas and we thought we were solving most of the problems we identified already.
+
+I read all of these issues and all the discussions I found and wrote this document that summarizes my proposal:
+*create a new notification system that we can customize and expand as we need in the future*:
+
+* A Django model to store the notifications' data
+* API endpoints to retrieve the notifications for a particular resource (User, Build, Project, Organization)
+* Frontend code to display them (*outside the scope of this document*)
+
+
+Goals
+-----
+
+* Keep raising exceptions for errors from the build process
+* Ability to add non-error notifications from the build process
+* Add extra metadata associated to the notification: icon, color, header, body, etc
+* Support different types of notifications (e.g. error, warning, note, tip)
+* Re-use the new notification system for product updates (e.g. new features, deprecated config keys)
+* Message content lives on Python classes that can be translated and formatted with objects (e.g. Build, Project)
+* Message could have richer content (e.g. HTML code) to generate links and emphasis
+* Notifications have trackable state (e.g. unread (default)=never shown, read=shown, dismissed=don't show again)
+* An object (e.g. Build, Organization) can have more than 1 notification attached
+* Remove hardcoded notifications from the templates
+* Notifications can be attached to Project, Organization, Build and User models
+* Specific notifications can be shown under the user's bell icon
+* Easy way to cleanup notification on status changes (e.g. subscription failure notification is auto-deleted after CC updated)
+* Notifications attached to Organization/Project dissappear for all the users once they are dismissed by anyone
+
+
+Non-goals
+---------
+
+* Create new Build "state" or "status" option for these fields
+* Implement the new notification in the old dashboard
+* Define front-end code implementation
+* Replace email or webhook notifications
+
+
+Small notes and other considerations
+------------------------------------
+
+* Django message system is not enough for this purpose.
+* Use a new model to store all the required data (expandable in the future)
+* How do we handle translations?
+  We should use ``_("This is the message shown to the user")`` in Python code and return the proper translation when they are read.
+* Reduce complexity on ``Build`` object (remove ``Build.status`` and ``Build.error`` fields among others).
+* Since the ``Build`` object could have more than 1 notification, when showing them, we will sort them by importane: errors, warnings, note, tip.
+* In case we need a pretty specific order, we can add an extra field for that, but it adds unnecessary complexity at this point.
+* For those notifications that are attached to the ``Project`` or ``Organization``, should it be shown to all the members even if they don't have admin permissions?
```

Comment:
> This might have to be done on a per-notification basis

Sounds complex to me. How would we define that on the notification?

I would start simple for now like "show to admin/owners" only for now and grow from there if we eventually need more granularity, probably.